pika/connection.py

```diff
@@ -1975,6 +1984,15 @@ def _on_connection_tune(self, method_frame):
         self.params.frame_max = Connection._negotiate_integer_value(self.params.frame_max,
                                                                     method_frame.method.frame_max)
 
+        if callable(self.params.heartbeat):
+            ret_heartbeat = self.params.heartbeat(method_frame.method.heartbeat)
+            if not isinstance(ret_heartbeat, numbers.Integral):
```

Comment:
I thought about cannot reusing the setter logic but now I do not want the callback to return another callable - so essentially it is not duplicating the same logic because the logic of the check changes when then TUNE connection state is reached. I thought about making the logic in heartbeat's setter in Parameters depend on the Connection state (ie. do not allow callable during TUNE state) but I thought doing so is not a clean solution (would then need to add another property inside Parameters for the connection state).

Reply:
Hehe ok ... still not completely untangled but for sure its better than my version :D